diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index 5822ee0..2cec9fa 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -1892,12 +1892,11 @@ static const struct file_operations btrfs_ctl_fops = {
 };
 
 static struct miscdevice btrfs_misc = {
-	.minor		= BTRFS_MINOR,
+	.minor		= MISC_DYNAMIC_MINOR,
 	.name		= "btrfs-control",
 	.fops		= &btrfs_ctl_fops
 };
 
-MODULE_ALIAS_MISCDEV(BTRFS_MINOR);
 MODULE_ALIAS("devname:btrfs-control");
 
 static int btrfs_interface_init(void)
-- 
1.8.3.1

