Index: linux-3.10.0-229.el7.centos.x86_64_btrfs_patch/fs/btrfs/extent_io.c
===================================================================
--- linux-3.10.0-229.el7.centos.x86_64_btrfs_patch.orig/fs/btrfs/extent_io.c
+++ linux-3.10.0-229.el7.centos.x86_64_btrfs_patch/fs/btrfs/extent_io.c
@@ -2755,9 +2755,7 @@ static int submit_extent_page(int rw, st
 {
 	int ret = 0;
 	struct bio *bio;
-	int nr;
 	int contig = 0;
-	int this_compressed = bio_flags & EXTENT_BIO_COMPRESSED;
 	int old_compressed = prev_bio_flags & EXTENT_BIO_COMPRESSED;
 	size_t page_size = min_t(size_t, size, PAGE_CACHE_SIZE);
 	struct btrfs_io_bio *io_bio;
@@ -2784,12 +2782,9 @@ static int submit_extent_page(int rw, st
 			return 0;
 		}
 	}
-	if (this_compressed)
-		nr = BIO_MAX_PAGES;
-	else
-		nr = bio_get_nr_vecs(bdev);
 
-	bio = btrfs_bio_alloc(bdev, sector, nr, GFP_NOFS | __GFP_HIGH);
+	bio = btrfs_bio_alloc(bdev, sector, BIO_MAX_PAGES,
+			      GFP_NOFS | __GFP_HIGH);
 	if (!bio)
 		return -ENOMEM;
 
