From 5378a2ac7342776ffe503bbfa589d590a924e82a Mon Sep 17 00:00:00 2001
From: Gu Zheng <gzheng@ddn.com>
Date: Tue, 1 Mar 2016 09:13:06 -0500
Subject: [PATCH 6/6] btrfs:add ifree support for btrfs

---
 fs/btrfs/super.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index bc2110a..231aa3b 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -1822,6 +1822,8 @@ static int btrfs_calc_avail_data_space(struct btrfs_root *root, u64 *free_bytes)
  * FIXME: not accurate for mixed block groups, total and free/used are ok,
  * available appears slightly larger.
  */
+#define MAX_INODE_SIZE	1092
+
 static int btrfs_statfs(struct dentry *dentry, struct kstatfs *buf)
 {
 	struct btrfs_fs_info *fs_info = btrfs_sb(dentry->d_sb);
@@ -1830,6 +1832,7 @@ static int btrfs_statfs(struct dentry *dentry, struct kstatfs *buf)
 	struct btrfs_space_info *found;
 	u64 total_used = 0;
 	u64 total_free_data = 0;
+	u64 total_free_metadata = 0;
 	int bits = dentry->d_sb->s_blocksize_bits;
 	__be32 *fsid = (__be32 *)fs_info->fsid;
 	unsigned factor = 1;
@@ -1860,6 +1863,11 @@ static int btrfs_statfs(struct dentry *dentry, struct kstatfs *buf)
 				}
 			}
 		}
+		if (found->flags & BTRFS_BLOCK_GROUP_METADATA) {
+			total_free_metadata += found->disk_total - found->disk_used;
+			total_free_metadata -=
+				btrfs_account_ro_block_groups_free_space(found);
+		}
 
 		total_used += found->disk_used;
 	}
@@ -1876,10 +1884,16 @@ static int btrfs_statfs(struct dentry *dentry, struct kstatfs *buf)
 	spin_unlock(&block_rsv->lock);
 
 	buf->f_bavail = div_u64(total_free_data, factor);
+	buf->f_ffree = total_free_metadata;
 	ret = btrfs_calc_avail_data_space(fs_info->tree_root, &total_free_data);
 	if (ret)
 		return ret;
 	buf->f_bavail += div_u64(total_free_data, factor);
+	buf->f_ffree += total_free_data;
+	if (buf->f_ffree < 40 * fs_info->fs_root->nodesize)
+		buf->f_ffree = 0;
+	else
+		buf->f_ffree /= MAX_INODE_SIZE;
 	buf->f_bavail = buf->f_bavail >> bits;
 
 	buf->f_type = BTRFS_SUPER_MAGIC;
-- 
1.8.3.1

